utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -0,0 +1,497 @@
+     */
+    private final AtomicLong outstandingDemand = new AtomicLong();
+
+    /**
+     * The queue of events to be processed, in the order they should be processed.
+     *
+     * <p>All logic within this publisher is represented using events in this queue. This ensures proper ordering of events
+     * processing and simplified reasoning about thread safety.
+     */
+    private final Queue<QueueEntry<T>> eventQueue = new ConcurrentLinkedQueue<>();
+
+    /**
+     * When processing the {@link #eventQueue}, these are the entries that should be skipped (and failed). This is used to
+     * safely drain the queue when there are urgent events needing processing, like a {@link Subscription#cancel()}.
+     */
+    private final Set<QueueEntry.Type> entryTypesToFail = new CopyOnWriteArraySet<>();
+
+    /**
+     * Whether the {@link #eventQueue} is currently being processed. Only one thread may read events from the queue at a time.
+     */
+    private final AtomicBoolean processingQueue = new AtomicBoolean(false);
+
+    /**
+     * An exception that should be raised to any failed {@link #write(Object)}, {@link #complete()} or {@link #error(Throwable)}
+     * operations. This is used to stop accepting messages after the downstream subscription is cancelled or after the
+     * caller sends a {@code complete()} or {@code #error()}.
+     *
+     * <p>This is a supplier to avoid the cost of creating an exception in the successful code path.
+     */
+    private final AtomicReference<Supplier<RuntimeException>> rejectException = new AtomicReference<>();
+
+    /**
+     * The subscriber provided via {@link #subscribe(Subscriber)}. This publisher only supports a single subscriber.
+     */
+    private Subscriber<? super T> subscriber;
+
+    /**
+     * Write a message to this publisher.
+     *
+     * <p>Messages written to this publisher will eventually be sent to a downstream subscriber, in the order they were
+     * written. When the message is sent to the subscriber, the returned future will be completed successfully.
+     *
+     * <p>This method may be invoked concurrently when the order of messages is not important.
+     *
+     * <p>In the time between when this method is invoked and the returned future is not completed, this publisher stores the
+     * request message in memory. Callers are recommended to limit the number of writes in progress at a time to bound the
+     * amount of memory used by this publisher.
+     *
+     * <p>The returned future will be completed exceptionally if the downstream subscriber cancels the subscription, or
+     * if the write call was performed after a {@link #complete()} or {@link #error(Throwable)} call.
+     *
+     * @param value The message to send. Must not be null.
+     * @return A future that is completed when the message is sent to the subscriber.
+     */
+    public CompletableFuture<Void> write(T value) {
+        log.trace(() -> "Received write() with " + value);
```

Comment:
We can't get a size since it's untyped within this publisher. We could log the type, but that's not as useful. Since it's at the trace level, and we already log sensitive data elsewhere at the debug level, shouldn't it be safe?

Reply:
Ah yeah, I just replaced T with ByteBuffer in my mind automatically somehow. Should we not include value at all then? My concern is that if it's a 8MB byte buffer, it's going to be a lot of data to be printed out.

> we already log sensitive data elsewhere at the debug level

IIRC, the logs that could contain potential sensitive data are wirelogs generated from HTTP clients, which is out of our control. Do we log those data elsewhere in the SDK?

Reply:
Logging a ByteBuffer doesn't log the contents, just the locations of the position, limit and capacity. The only chance at logging sensitive data is if we reused this elsewhere for things that do have a sensitive toString.

Reply:
Ah I see, then I think it should be fine